Fix race condition in KubernetesExecutor with concurrently running schedulers

droppoint opened the pull request:
Closes: #32928

A race condition occurs in the _adopt_completed_pods function when schedulers are running concurrently. _adopt_completed_pods function doesn't keep track of which scheduler went down so it constantly tries to adopt completed pods from normally working schedulers. On Airflow setups with concurrently running schedulers and with a lot of short living DAG's it leads to race condition and open slots leak. You can find detailed analysis of this situation in GitHub issue here (https://github.com/apache/airflow/issues/32928#issuecomment-1820413530) Even if one of the schedulers went down and left some pods in completed state it is not a big deal because: 
1. Completed pods do not consume resources from kubernetes cluster 
2. They can be deleted by airflow cleanup-pods CLI command which you can execute with the cronjob.

Co-authored-by: Vlad Pastushenko iam@vladpi.me

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->

droppoint committed:
Fix race condition in KubernetesExecutor with concurrently running schedulers

A race condition occurs in the _adopt_completed_pods function when schedulers are running concurrently. _adopt_completed_pods function doesn't keep track of which scheduler went down so it constantly tries to adopt completed pods from normally working schedulers. On Airflow setups with concurrently running schedulers and with a lot of short living DAG's it leads to race condition and open slots leak. You can find detailed analysis of this situation in GitHub issue here (https://github.com/apache/airflow/issues/32928#issuecomment-1820413530) Even if one of the schedulers went down and left some pods in completed state it is not a big deal because: 1. Completed pods do not consume resources from kubernetes cluster 2. They can be deleted by airflow cleanup-pods CLI command which you can execute with the cronjob.

Co-authored-by: Vlad Pastushenko iam@vladpi.me

JCoder01 commented on airflow/providers/cncf/kubernetes/executors/kubernetes_executor.py:
```diff
@@ -642,39 +641,6 @@ def adopt_launched_task(
-                    body={"metadata": {"labels": {"airflow-worker": new_worker_id_label}}},
-                )
-            except ApiException as e:
-                self.log.info("Failed to adopt pod %s. Reason: %s", pod.metadata.name, e)
```
Rather than deleting all this functionality, couldn't you just drop a 'continue' on line 675 so that if the patch fails it doesn't add the task to 'self'?

ephraimbuddy replied:
This should work and I feel it's better. 675 & 676

droppoint replied:
It's not that simple. While it's true that placing "continue" in the except block will fix part of the problem, it won't resolve the race condition. Slots will continue to leak after that. Please read this [comment](https://github.com/apache/airflow/issues/32928#issuecomment-1820413530) from issue #32928. In short, the race happens because this function attempts to adopt completed pods every 5 minutes (by default) from other fully operational schedulers. When this 'adoption' occurs, the TaskInstance key of normally completed pod is added to KubernetesExecutor.running set. Since both the pod and TaskInstance are complete, no change_state will occur, leaving the TaskInstance key in the KubernetesExecutor.running set indefinitely (or until the scheduler is restarted). Once the count of KubernetesExecutor.running exceeds KubernetesExecutor.parallelism, the executor will cease scheduling new tasks.

JCoder01 replied:
Apologies. I had read your comment but due to the hazards of reading too fast, too early in the morning I missed that you had already tried the `continue` approach. Reviewing your comment again, it seems adding the `continue` solves the majority of the issues. This leaves the remaining condition which if I understand it, is that it's possible for 2 schedulers to adopt a completed pod, which means it's added to the `running` set. And then once the pod is deleted by one scheduler there is nothing to signal to the other scheduler(s) that the pod needs to be removed from the `running` set? So we need to either :
A. only adopt pods not owned by a running executor, or 
B. Clean up all schedulers when a pod is deleted. 
The proper approach seems to be what @dstandish recommended to check if the pod is owned by a live scheduler before adopting. Or alternatively, perhaps KubernetesJobWatcher to could react to `DELETE` events rather than ignoring them, but this seems far less desirable to just not adopting the pods to start with if running multiple schedulers.
The CronJob approach just seems like the wrong direction. I don't think the option to create it as part of the chart exists in charts other than the official chart, and even so, in the official chart it's disabled by default.

droppoint replied:
We chose the approach "A" and refactored the function. https://github.com/apache/airflow/pull/35800#issuecomment-1845545959

droppoint committed:
Fix race condition in KubernetesExecutor with concurrently running schedulers

A race condition occurs in the _adopt_completed_pods function when schedulers are running concurrently. _adopt_completed_pods function doesn't keep track of which scheduler went down so it constantly tries to adopt completed pods from normally working schedulers. On Airflow setups with concurrently running schedulers and with a lot of short living DAG's it leads to race condition and open slots leak. You can find detailed analysis of this situation in GitHub issue here (apache#32928 (comment)). The _adopt_completed_pods function was refactored to the _delete_orphaned_completed_pods function, which removes only completed pods that are not bound to running schedulers.

Co-authored-by: Vlad Pastushenko <iam@vladpi.me>

potiuk reviewed (approved):
It LGTM ... But I 'd  love 2nd/3rd opinion.

dirrao commented on airflow/providers/cncf/kubernetes/executors/kubernetes_executor.py:
```diff
@@ -642,38 +642,37 @@ def adopt_launched_task(
+        ).all()
+        labels = ["kubernetes_executor=True", f"{POD_EXECUTOR_DONE_KEY}!=True"]
+        for alive_scheduler_id in alive_schedulers_ids:
+            labels.append(f"airflow-worker!={self._make_safe_label_value(str(alive_scheduler_id))}")
```
Let deletion of completed pods should be taken care by each scheduler (distribute the pod deletion work).

droppoint replied:
I'm afraid I don't fully understand what are you suggesting. Can you please explain it more?

If your concern is about performance, then I think you shouldn't be concerned at all. The situation where a completed pod becomes an orphan is pretty rare. It only occurs in Airflow setups that fulfill two conditions: 1) One of the schedulers died, and 2) Your Airflow setup has a lot of simultaneously working dags/pods. The time window when a pod becomes completed and before it's normally deleted is quite small. My Airflow setup has about 2k simultaneously working dags, but I have only several completed pods that were orphaned after one of the schedulers was down. You can make delete calls for several pods quickly.

In addition to that, I don't see how we can distribute the deletion of orphaned completed pods between the rest of the schedulers. To find which pods should be deleted, you need to send a list_pods request through the Kubernetes client and supply the list of all ids of the schedulers that are still alive. You can't submit a part of the list, as it would lead to the deletion of completed pods from normally working schedulers.

dirrao replied:
Instead of going through the all alive schedulers, why can't we just delete completed pods from other schedulers rather its own. If the completed pod is owned itself, then that we will taken care through wither the success event or other scheduler.

droppoint replied:
Because in this case, you will delete/adopt completed pods from other normally working schedulers. For example, consider a setup with 3 schedulers: A, B, and C. Suppose scheduler A is down. When the 'adoption' procedure starts in either scheduler B or scheduler C, they should adopt running and completed pods from scheduler A but not from each other since they are working normally. So if the label select has a label like 'airflow-worker!=<current scheduler id>', you will delete/adopt pods that are not meant for adoption.

potiuk replied:
Does it matter which scheduler deletes completed pods? I think this is the last stage of a POD and whoever deletes it, is doing a good thing. I believe the other schedulers will simply ignore if they see such completed PODs  have been deleted, so it does not matter which of the A, B, C schedulers deletes such a POD IMHO. - and that also gives the opportunity of cleaning them up faster, as the three scheduling loops will generally happen 3 times more often and completed pods will be deleted more frequently (which is a good thing).

IMHO there is no particular reason to tie completed POD deletion to the scheduler that created it. But maybe there is a reason and I miss it?

dirrao replied:
Yes. I don't see any issue with performance. However, I see one problem mentioned in the below thread related to the executor open slots leak.

dirrao reviewed (commented):
Thanks for MR. I have given one comment.

dirrao commented on airflow/providers/cncf/kubernetes/executors/kubernetes_executor.py:
```diff
@@ -642,38 +642,37 @@ def adopt_launched_task(
             except ApiException as e:
-                self.log.info("Failed to adopt pod %s. Reason: %s", pod.metadata.name, e)
-            ti_id = annotations_to_key(pod.metadata.annotations)
-            self.running.add(ti_id)
```
We need to gracefully free the executor running slot from the concerned scheduler. Right now, this happens through the pod adoption of completed pods. 

```
        if event["type"] == "DELETED" and not pod.metadata.deletion_timestamp:
            # This will happen only when the task pods are adopted by another executor.
            # So, there is no change in the pod state.
            # However, need to free the executor slot from the current executor.
            self.log.info("Event: pod %s adopted, annotations: %s", pod_name, annotations_string)
            self.watcher_queue.put((pod_name, namespace, ADOPTED, annotations, resource_version))
        elif status == "Succeeded":
            # We get multiple events once the pod hits a terminal state, and we only want to
            # send it along to the scheduler once.
            # If our event type is DELETED, we have the POD_EXECUTOR_DONE_KEY, or the pod has
            # a deletion timestamp, we've already seen the initial Succeeded event and sent it
            # along to the scheduler.
            if (
                event["type"] == "DELETED"
                or POD_EXECUTOR_DONE_KEY in pod.metadata.labels
                or pod.metadata.deletion_timestamp
            ):
                self.log.info(
                    "Skipping event for Succeeded pod %s - event for this pod already sent to executor",
                    pod_name,
                )
                return
            self.log.info("Event: %s Succeeded, annotations: %s", pod_name, annotations_string)
            self.watcher_queue.put((pod_name, namespace, None, annotations, resource_version))
```

droppoint replied:
Can you clarify what you mean by a 'concerned scheduler'? I understand that _adopt_completed_pods function is needed when one of the schedulers is down (i.e., killed by OOM). When your pod is completed, and if the scheduler that it was attached to is down, the TaskInstance corresponding to this pod is not in any running set of the rest of the schedulers. And you don't have to adopt it (or add it to the running set of any alive scheduler) because it's already been completed. Perhaps this comment will provide more information for you: link (https://github.com/apache/airflow/pull/35800#issuecomment-1845545959).

Are you referring to the situation when the scheduler missed its heartbeat and was announced as 'failed' by another scheduler? Shouldn't we fail loudly in this situation? Should we really mask the problem with skipping heartbeats this way?

dirrao replied:
Yes. I am referring to the missing heartbeat scenario. This will cause the executor running slot leak due to its been deleted by other scheduler due to its current scheduler heartbeat is delayed.

droppoint replied:
Ok we'll try to refactor this function to it's initial state + change kube selector to adopt pods only from failed scheduler + add continue to the except block. We will run this patch on our airflow setup and will comeback with the results.

dirrao replied:
The original adoption code will adopt the pods from other schedulers including the failed ones. Once they are adopted. the watcher will get a successful event. and then the executor gracefully deletes the pod including freeing the running queue slot.  I would request you to try the latest main code to see if that fixes this problem. The latest main code has executor running slot leak fix.

droppoint replied:
Adoption should be performed only on pods that are bound to failed schedulers. It should not be performed on pods that are bound to normally working schedulers.

potiuk replied:
> Adoption should be performed only on pods that are bound to failed schedulers. It should not be performed on pods that are bound to normally working schedulers.

Why would that apply to completed PODs as well? What problem would it cause? 

I understand that what we are trying to do here is not to `adopt` those pods, but to `delete`  completed ones (this is the gist of that change as I see it). There is a bit of difference in `adopting` and taking care about the POD, than `deleting` the POD (which is the last thing that happens to it). As I see see it, yes, adoption means "I will take care about this POD in the future", where "deletion" means "let's delete this pod and not care about it any more". Those are two completely differnt cases.

Or am I missing something?

dirrao replied:
Let's say we have a scenario as follows. Scheduler 1 added the task to the executor running queue, launched the worker pod, and then succeeded. As per the proposed code, Let's say scheduler 2 deleted this completed pod, then the scheduler1  executor running queue has a task entry and it will be there forever. It is a K8 executor running queue slot leak. As per the existing code, Let's say the scheduler 2 adopted this completed pod. The DELETE event is captured in schedueler1 watcher and then it frees the executor running queue slot. The scheduler2 will get an ADDED event followed by a SUCCEDED event and then the completed pod will be deleted. 
https://github.com/apache/airflow/blob/bf90992dd48bce7de9f2a687860479e95575cd24/airflow/providers/cncf/kubernetes/executors/kubernetes_executor_utils.py#L245C5-L245C5

droppoint replied:
I think I misinterpreted "missed heartbeat" situation. I need a little more time to properly think about it.

dstandish replied:
Not trying to muddy the waters here @dirrao but I am curious about your comment that scheduler 1 gets DELETED event when pod adopted.... if the pod were actually deleted, wouldn't there also be a DELETED event? (though for a different reason)

Also, there's something messed up in the main code that you quoted earlier in this thread.  you have logic like this
```
if event == deleted:
    ...
elif ...:
    if event == deleted
```

that second if event == deleted will never be reached.

dirrao replied:
Right now, the Airflow K8s watcher is subscribed to pod events with a scheduler id label selector. When the scheduler ID is changed in pod labels during adoption of queued tasks or completed tasks, the original scheduler gets a DELETE event but it's not deleted. The way to identify the deletion of a pod is based on **pod.metadata.deletion_timestamp**. 

```
        if event["type"] == "DELETED" and not pod.metadata.deletion_timestamp:
               print("pod is adopted")
        elif event["type"] == "DELETED" and pod.metadata.deletion_timestamp:
               print("pod is deleted")
```

hussein-awala commented on airflow/providers/cncf/kubernetes/executors/kubernetes_executor.py:
```diff
@@ -642,38 +642,37 @@ def adopt_launched_task(
+        for alive_scheduler_id in alive_schedulers_ids:
+            labels.append(f"airflow-worker!={self._make_safe_label_value(str(alive_scheduler_id))}")
+
+        query_kwargs = {"field_selector": "status.phase=Succeeded", "label_selector": ",".join(labels)}
```
I think we need to delete failed pods too if `delete_worker_pods_on_failure` is `True`.

hussein-awala reviewed (commented):
Great work; it looks good.

I just added a small comment about deleting failed workers when `delete_worker_pods_on_failure` is set to `True` to avoid having ophan failed pods when we lose the scheduler.

dstandish commented on airflow/providers/cncf/kubernetes/executors/kubernetes_executor.py:
```diff
@@ -642,38 +642,37 @@ def adopt_launched_task(
+            select(Job.id).where(Job.job_type == "SchedulerJob", Job.state == JobState.RUNNING)
+        ).all()
+        labels = ["kubernetes_executor=True", f"{POD_EXECUTOR_DONE_KEY}!=True"]
+        for alive_scheduler_id in alive_schedulers_ids:
+            labels.append(f"airflow-worker!={self._make_safe_label_value(str(alive_scheduler_id))}")
```
It looks like you can do `notin` https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#set-based-requirement
e.g. `airflow-worker notin (1,2,3)`

dstandish commented on airflow/providers/cncf/kubernetes/executors/kubernetes_executor.py:
```diff
@@ -642,38 +642,37 @@ def adopt_launched_task(
+            labels.append(f"airflow-worker!={self._make_safe_label_value(str(alive_scheduler_id))}")
+
+        query_kwargs = {"field_selector": "status.phase=Succeeded", "label_selector": ",".join(labels)}
         pod_list = self._list_pods(query_kwargs)
```
if the only alive scheduler is _this_ scheduler, we can return right here

dstandish reviewed (commented):
a suggestion or two.  @hussein-awala's suggestion looks correct to me. with these, this should be good to merge.
